### ISSUE TITLE
Feature: Implement DNA and Genes Model for Agent Parameters in Simulation

### ISSUE THREAD

**Issue by csmangum** [Dooders/Experiments#22]
This issue aims to model simulation agent parameters using a DNA and genes-inspired approach. By structuring agent parameters, such as “learning rate,” as genes within a genome, we can encode, decode, store, and evolve them through genetic algorithm-inspired processes. This approach will involve creating representations similar to DNA (chromosomes with gene sequences), where parameters are encoded, manipulated, and decoded to control agent behavior within the simulation.

---

**Objectives**

1. Design a DNA-inspired structure for encoding agent parameters as chromosomes.
2. Define encoding and decoding functions for translating parameters between binary or real-valued gene representations and functional values.
3. Implement genetic processes (selection, crossover, mutation) to evolve these parameters over generations.
4. Experiment with evolving learning rates using this genetic encoding to observe convergence patterns.
5. Document and analyze the results, focusing on the effectiveness and interpretability of the DNA and gene model for agent parameters.

---

**Key Tasks**

### 1. **Define DNA Representation for Agent Parameters**

   - **Chromosome Design**: Create a chromosome structure (e.g., a list or array) to hold all agent parameters. Each parameter (gene) will represent a specific trait, such as learning rate.
   - **Gene Definition**: Define individual genes within the chromosome, with each gene corresponding to an encoded parameter value. For example:
      - Learning Rate Gene
      - Other potential genes if applicable (e.g., exploration factor, memory size).
   - **Alleles**: Specify ranges and types for gene values (e.g., binary, real-valued).

   **Example**:
   ```python
   # Chromosome Example: [Gene1 (learning rate), Gene2, ...]
   chromosome = [0.05, 0.8, ...]
   ```

### 2. **Encode Parameters into Genes**

   - **Binary Encoding**:
     - If using binary encoding, convert parameter values to binary form.
     - Map binary strings to meaningful parameter ranges.
     - Example: Convert `learning rate = 0.05` into an 8-bit binary string.
   - **Real-Valued Encoding**:
     - Directly encode real values within defined ranges.
     - Normalize values to simplify manipulation.
   
   **Example**:
   ```python
   # Encoding example for learning rate in range [0.001, 0.1]
   learning_rate_gene = int((learning_rate - 0.001) / (0.1 - 0.001) * 255)
   ```

### 3. **Implement Decoding Function for Genes**

   - Decode binary or real-valued genes back into usable parameters.
   - Ensure proper mapping from genetic representation to functional values.

   **Example**:
   ```python
   # Decoding example for learning rate gene
   decoded_learning_rate = 0.001 + (learning_rate_gene / 255) * (0.1 - 0.001)
   ```

### 4. **Create Genetic Operators for Evolution**

   - **Selection**: Select agent genomes based on fitness scores.
      - Example methods: Roulette wheel selection, tournament selection.
   - **Crossover**: Implement methods to combine genes from parent chromosomes.
      - Single-point, two-point, or uniform crossover methods.
   - **Mutation**: Apply mutations to introduce variations in genes.
      - Binary mutation (flip bits) or real-valued mutation (add small random values).
   
   **Example**:
   ```python
   def crossover(parent1, parent2):
       # Single-point crossover for simplicity
       crossover_point = random.randint(0, len(parent1) - 1)
       return parent1[:crossover_point] + parent2[crossover_point:]
   ```

### 5. **Integrate Evolutionary Process in Simulation Loop**

   - **Initialization**: Generate initial population with randomized chromosomes.
   - **Fitness Evaluation**: Evaluate each agent’s fitness based on parameters like learning rate.
   - **Evolution Cycle**:
      - Perform selection, crossover, and mutation.
      - Generate new population and repeat for multiple generations.
   - **Logging**: Track gene evolution and parameter convergence over generations.

### 6. **Run Experiments and Analyze**

   - Conduct experiments across multiple generations with varied mutation rates, selection pressures, and crossover techniques.
   - Record the evolution of learning rates and other parameters over time.
   - Visualize convergence patterns and analyze how effectively the genetic model optimizes parameters.

---

**Acceptance Criteria**

- [ ] Agent parameters are encoded as genes in a chromosome-like structure.
- [ ] Encoding and decoding functions are implemented for gene-to-parameter translation.
- [ ] Genetic operators (selection, crossover, mutation) are functional for evolving agent genomes.
- [ ] Evolutionary process successfully runs over multiple generations, with learning rate convergence tracked and logged.
- [ ] Experiment results are analyzed, with findings documented.

---

**Additional Notes**

- **Visualization**: Consider visualizing parameter evolution (e.g., average learning rate per generation).
- **Extensions**: If the model works well, apply this DNA structure to more agent parameters.
- **Performance**: Monitor the efficiency of encoding/decoding and consider optimization for larger populations.

**References**

- Documentation on genetic algorithms and genetic programming.
- Background on biological encoding models for inspiration.

---

### Labels

- `feature-request`
- `enhancement`
- `experiment`
- `genetic-algorithm`
- `biological-model`

**Comment by csmangum** [comment]
1. **Chromosome Structure**:
   - Think of each agent’s genome as a chromosome containing genes.
   - Each gene represents a specific parameter (e.g., learning rate) or trait of the agent.
   - The chromosome can be a list or array with values encoding these parameters.

2. **Encoding Genes**:
   - Use **binary encoding** (e.g., 8-bit or 16-bit) if you want the parameters to be highly discrete. Binary encoding is especially useful for mutation operations, allowing precise bit-level changes.
   - Alternatively, use **real-valued encoding** if you prefer a continuous representation for parameters like learning rate. This approach keeps the values in the desired range and provides smoother adjustments when applying mutation.

3. **Mapping Parameter Ranges**:
   - Define a range for each parameter. For instance, map the learning rate between 0.001 and 0.1.
   - If using binary encoding, determine the number of bits to represent the resolution within this range (e.g., 8 bits allow 256 distinct values).

4. **Genetic Operators**:
   - **Selection**: Select agents based on a fitness function that evaluates the performance associated with the encoded parameters (e.g., how effectively agents learn using their learning rate).
   - **Crossover**: Implement crossover to combine genes of two parent agents. A single-point crossover is straightforward, swapping genes from a randomly chosen point to produce offspring.
   - **Mutation**: Randomly alter genes to maintain diversity. In binary encoding, mutate by flipping bits; for real-valued encoding, add small random changes.

5. **Evolutionary Cycle**:
   - Each generation consists of fitness evaluation, selection, crossover, mutation, and replacement.
   - Start with a random population of agents, evolve over multiple generations, and observe if parameters like learning rate converge toward optimal values.
   - Track parameter evolution by logging average values per generation, allowing you to analyze convergence trends.

6. **Fitness Function**:
   - Define a fitness function that uses the learning rate to assess agent performance. For instance, evaluate how quickly agents reach a target accuracy or optimal performance metric.
   - Make the fitness function sensitive enough to reward learning rate values that balance stability and speed in the learning process.

7. **Logging and Visualization**:
   - Track parameter evolution by logging each generation’s average, minimum, and maximum values.
   - Plot the convergence of learning rate over generations to visualize how the population adapts.

**Comment by csmangum** [comment]
## Sub-issues (gap breakdown for this epic)

Tracked work split from the gap analysis:

| # | Topic |
|---|--------|
| #676 | Chromosome / gene schema for tunable hyperparameters |
| #677 | Encode/decode (gene representation ↔ values) |
| #678 | Selection operators (tournament, roulette) |
| #679 | Crossover + mutation on hyperparameter gene vectors |
| #680 | Simulation / experiment runner integration (multi-generation GA) |
| #681 | Logging, visualization, and documentation of convergence |

Suggested order: **676 → 677 → 678 + 679** (can parallelize operators) **→ 680 → 681**.

**Comment by csmangum** [comment]
The gene-based evolution experiments show a clear pattern:

- **The genetics pipeline works end-to-end**: chromosome encoding/decoding, crossover, mutation, selection, lineage logging, and per-generation stats are all functioning.
- **`learning_rate` is highly influential but unstable**: it often drives large fitness changes and can collapse to the lower bound (`1e-6`) under some settings.
- **Selection/mutation trade-off is real**:
  - **Tournament + lower mutation** tends to produce stronger optimization signal.
  - **Roulette + higher mutation** keeps diversity but often stalls fitness improvement.
- **Boundary handling matters a lot**:
  - **Reflect mode** was the most reliable anti-collapse strategy in your runs (no lower-bound hits in the documented comparison and competitive/better fitness).
  - **Clamp mode** frequently showed boundary collapse (many candidates at `1e-6`).
- **Soft boundary penalties help, but not consistently**:
  - Penalty strength effects were **non-monotonic** (e.g., `0.02` and `0.10` helped in some runs, `0.05` still collapsed late).
- **Multi-gene tracking is now in place**:
  - You’re no longer just watching `learning_rate`; summaries track `gamma`, `epsilon_decay`, and `memory_size` too.

Practical takeaway: the gene model is valid and useful, but convergence quality depends heavily on mutation pressure + boundary strategy. The best next default is **tournament selection + reflective boundaries + moderate mutation**.
